core/include/vecmem/containers/details/vector_buffer.hpp

```diff
@@ -39,7 +39,7 @@ namespace vecmem::details {
 
    private:
       /// Data object owning the allocated memory
-      std::unique_ptr< TYPE, deleter > m_memory;
+      std::unique_ptr< TYPE, deallocator > m_memory;
```

Comment:
Why are using `deallocator` here and not `deleter`? If I understand correctly, `deleter` is now the class that actually does the destruction that you would expect it to do, right?

Reply:
It's the same issue that we had with the "old design", and why the memory managers needed to declare whether they provide host-accessible or non-host-accessible memory.

`vecmem::details::vector_buffer` is mainly meant to be used with non-host-accessible memory. As such, `vecmem::details::deleter` would try to access memory from the host that it cannot.

I myself imagine the following things here:
  - For CUDA/HIP/SYCL "managed" memory we can let the allocation/deallocation call the constructor/destructor of non-trivial types as it wants. For non-host-accessible memory we can't do either, we can just allocate and deallocate the memory needed for those types.
  - For "simple" types, which is mostly what we should be using here anyway, this should be enough as-is.
  - For "complex" types this does put a limitation on what we can and cannot do.

In general I do not believe that any "GPU friendly" data type should ever do anything important in its destructor. These data types are not allowed to manage their own memory. So their destructor should always be a default one. (We could very well add a `static_assert` check for that...)

Having non-default constructors is another matter however. As those will be used for any operations that need to set the memory describing the object "to a ground/default state".

Now... For device-only memory, for input variables, this is not a problem. There you first need to create the objects in host memory, using their constructors. Then the memory used by these host-created objects can be copies verbatim to the device using one of the copy functions that we'll be adding in addition to the CUDA ones. After this, the objects created around that memory block in the device code should see the objects in a "reasonable state".

However for output variables this may prove a bit more tricky. Since the code currently only allocates memory for the output variables. It doesn't initialise that variable in any shape or form. The `vecmem::device_vector` and `vecmem::device_array` types would need to learn how to use the constructors of the types that they describe to initialise the memory given to them.

But... Very long story short, the `vecmem::details::vector_buffer` type, in my mind, must only do memory allocations and de-allocations. It must not interact with the types that would eventually use these allocated memory blocks...